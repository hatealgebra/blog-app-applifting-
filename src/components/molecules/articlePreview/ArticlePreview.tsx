import React from 'react';

import StyledArticlePreviewContainer, {
  StyledArticlePreviewImage,
  StyledArticleRow,
} from './articlePreview.styled';

import { GoPrimitiveDot } from '@react-icons/all-files/go/GoPrimitiveDot';
import StyledLink from '../../atoms/links/link.styled';
import { components } from '../../../types/declarations';

import { getDate } from '../../../utils/date.utils';

import {
  createArticleLink,
  cutTextWithElipsis,
} from '../../../utils/generic.utils';
import NoImageAvailable from '../noImageAvalaible/NoImageAvailable';

// TODO: Implement fallback image
const ArticlePreview = ({
  articleId,
  title,
  createdAt,
  perex,
  comments,
  imageBase64,
  author,
}: components['schemas']['ArticleDetail'] & {
  author: string;
  imageBase64: String;
}) => {
  const [fallbackImage, setFallbackImage] = React.useState(false);
  const createdDate = getDate(createdAt);
  const file = 'data:image/png;base64,' + imageBase64;
  const commentsArray = JSON.parse(comments);

  const useFallbackImage = ({ currentTarget }: Event) => {
    currentTarget.onerror = null;
    setFallbackImage(true);
  };

  return (
    <StyledArticlePreviewContainer className="article-preview">
      {fallbackImage ? (
        <NoImageAvailable />
      ) : (
        <StyledArticlePreviewImage
          className="article-preview__img"
          src={file}
          alt={`${title} preview image`}
          onError={({ currentTarget }) => useFallbackImage(currentTarget)}
        />
      )}
      <h3 className="article-preview__title">
        {cutTextWithElipsis(title, 50)}
      </h3>
      <StyledArticleRow className="article-preview__row-one">
        <span>{author ?? 'Unknown author'}</span>
        <GoPrimitiveDot />
        <span>{createdDate}</span>
      </StyledArticleRow>
      <p className="article-preview__text">{perex}</p>
      <StyledArticleRow className="article-preview__row-two">
        <StyledLink to={createArticleLink(articleId)}>
          Read whole article
        </StyledLink>
        <span>{`${commentsArray?.length || '0'} comments`}</span>
      </StyledArticleRow>
    </StyledArticlePreviewContainer>
  );
};

export default ArticlePreview;
