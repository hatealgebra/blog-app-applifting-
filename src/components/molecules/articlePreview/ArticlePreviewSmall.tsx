import React from 'react';
import {
  createArticleLink,
  cutTextWithElipsis,
} from '../../../utils/generic.utils';
import { StyledArticlePreviewSmall } from './articlePreview.styled';

const ArticlePreviewSmall = ({
  heading,
  children,
  articleId,
}: ArticlePreviewSmallProps) => {
  return (
    <StyledArticlePreviewSmall to={createArticleLink(articleId)}>
      <h5>{heading}</h5>
      <p>{cutTextWithElipsis(children, 220)}</p>
    </StyledArticlePreviewSmall>
  );
};

interface ArticlePreviewSmallProps {
  heading: string;
  children: string;
  articleId: string;
}

export default ArticlePreviewSmall;
