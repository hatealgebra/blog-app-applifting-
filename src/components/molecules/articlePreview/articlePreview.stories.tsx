import React from 'react';

import ArticlePreview from './ArticlePreview';
import ArticlePreviewSmall from './ArticlePreviewSmall';
import articlesDetailsResponseMockJSON from '../../../__mocks__/asyncData/get/articlesDetailsResponse.mock.json';

const { comments } = articlesDetailsResponseMockJSON.items[0];

export const ArticlePreviewMobile = () => {
  return (
    <ArticlePreview
      imageBase64={''}
      title="Article heading"
      author="Random Author"
      createdAt="08/08/2021"
      perex="Lorem ipsum dolor sit amet consectetur, adipisicing elit. Porro beatae quod qui id dolorum nam, sapiente rem debitis ad illo doloribus ab! Dolorem aliquam facilis labore fugit rem, dolores quasi."
      comments={comments}
    />
  );
};
export const ArticlePreviewLongTitle = () => (
  <ArticlePreview
    imageBase64={''}
    title="Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum"
    author="Random Author"
    createdAt="08/08/2021"
    perex="Lorem ipsum dolor sit amet consectetur, adipisicing elit. Porro beatae quod qui id dolorum nam, sapiente rem debitis ad illo doloribus ab! Dolorem aliquam facilis labore fugit rem, dolores quasi."
    comments={comments}
  />
);

export const ArticlePreviewSmallExample = () => (
  <ArticlePreviewSmall articleId="" heading="Hello world">
    Hello there
  </ArticlePreviewSmall>
);

export default {
  title: 'Molecules/Article Preview',
  component: ArticlePreview,
};
