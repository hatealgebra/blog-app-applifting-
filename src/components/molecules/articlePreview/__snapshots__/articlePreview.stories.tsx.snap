// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/Article Preview ArticlePreviewLongTitle smoke-test 1`] = `
<article class="sc-eLSjS kPKThh article-preview">
  <div class="sc-fnOeiS iGypIr">
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 16 16"
         color="white"
         height="50%"
         width="50%"
         xmlns="http://www.w3.org/2000/svg"
         style="color: white;"
    >
      <path fill-rule="evenodd"
            d="M14.002 2h-12a1 1 0 00-1 1v10a1 1 0 001 1h12a1 1 0 001-1V3a1 1 0 00-1-1zm-12-1a2 2 0 00-2 2v10a2 2 0 002 2h12a2 2 0 002-2V3a2 2 0 00-2-2h-12z"
            clip-rule="evenodd"
      >
      </path>
      <path d="M10.648 7.646a.5.5 0 01.577-.093L15.002 9.5V14h-14v-2l2.646-2.354a.5.5 0 01.63-.062l2.66 1.773 3.71-3.71z">
      </path>
      <path fill-rule="evenodd"
            d="M4.502 7a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
            clip-rule="evenodd"
      >
      </path>
    </svg>
  </div>
  <h3 class="article-preview__title">
    Lorem ipsum Lorem ipsum Lorem ipsum Lorem ipsum Lo...
  </h3>
  <div class="sc-ksJisA icZPyt article-preview__row-one">
    <span>
      Random Author
    </span>
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 8 16"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill-rule="evenodd"
            d="M0 8c0-2.2 1.8-4 4-4s4 1.8 4 4-1.8 4-4 4-4-1.8-4-4z"
      >
      </path>
    </svg>
    <span>
      07/00/21
    </span>
  </div>
  <p class="article-preview__text">
    Lorem ipsum dolor sit amet consectetur, adipisicing elit. Porro beatae quod qui id dolorum nam, sapiente rem debitis ad illo doloribus ab! Dolorem aliquam facilis labore fugit rem, dolores quasi.
  </p>
  <div class="sc-ksJisA icZPyt article-preview__row-two">
    <a class="sc-jrkPcm eEbEbs"
       href="/articles/undefined"
    >
      Read whole article
    </a>
    <span>
      4 comments
    </span>
  </div>
</article>
`;

exports[`Molecules/Article Preview ArticlePreviewMobile smoke-test 1`] = `
<article class="sc-eLSjS kPKThh article-preview">
  <div class="sc-fnOeiS iGypIr">
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 16 16"
         color="white"
         height="50%"
         width="50%"
         xmlns="http://www.w3.org/2000/svg"
         style="color: white;"
    >
      <path fill-rule="evenodd"
            d="M14.002 2h-12a1 1 0 00-1 1v10a1 1 0 001 1h12a1 1 0 001-1V3a1 1 0 00-1-1zm-12-1a2 2 0 00-2 2v10a2 2 0 002 2h12a2 2 0 002-2V3a2 2 0 00-2-2h-12z"
            clip-rule="evenodd"
      >
      </path>
      <path d="M10.648 7.646a.5.5 0 01.577-.093L15.002 9.5V14h-14v-2l2.646-2.354a.5.5 0 01.63-.062l2.66 1.773 3.71-3.71z">
      </path>
      <path fill-rule="evenodd"
            d="M4.502 7a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
            clip-rule="evenodd"
      >
      </path>
    </svg>
  </div>
  <h3 class="article-preview__title">
    Article heading
  </h3>
  <div class="sc-ksJisA icZPyt article-preview__row-one">
    <span>
      Random Author
    </span>
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 8 16"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill-rule="evenodd"
            d="M0 8c0-2.2 1.8-4 4-4s4 1.8 4 4-1.8 4-4 4-4-1.8-4-4z"
      >
      </path>
    </svg>
    <span>
      07/00/21
    </span>
  </div>
  <p class="article-preview__text">
    Lorem ipsum dolor sit amet consectetur, adipisicing elit. Porro beatae quod qui id dolorum nam, sapiente rem debitis ad illo doloribus ab! Dolorem aliquam facilis labore fugit rem, dolores quasi.
  </p>
  <div class="sc-ksJisA icZPyt article-preview__row-two">
    <a class="sc-jrkPcm eEbEbs"
       href="/articles/undefined"
    >
      Read whole article
    </a>
    <span>
      4 comments
    </span>
  </div>
</article>
`;

exports[`Molecules/Article Preview ArticlePreviewSmallExample smoke-test 1`] = `
<a class="sc-hBpgZr cFkYUG"
   href="/articles/"
>
  <h5>
    Hello world
  </h5>
  <p>
    Hello there
  </p>
</a>
`;
